core/src/main/java/io/spine/core/EventMixin.java

```diff
@@ -59,10 +59,11 @@ default Timestamp time() {
 
     @Override
     default MessageId rootMessage() {
-        EventContext.OriginCase origin = context().getOriginCase();
-        return origin == PAST_MESSAGE
-               ? context().getPastMessage().root()
-               : messageId();
+        return isImported()
+               ? messageId()
+               : context()
+                       .rootMessage()
+                       .orElseThrow(IllegalStateException::new);
```

Comment:
Are you sure it's a good idea to throw an exception here? I'd better return `Optional`. It's pretty uncommon to throw an exception from Protobuf message.